optimize/global.go

```diff
@@ -10,106 +10,74 @@ import (
 	"time"
 )
 
-var (
-	nonpositiveDimension string = "optimize: non-positive input dimension"
-	negativeTasks        string = "optimize: negative input number of tasks"
-)
+// DefaultSettingsGlobal returns the default settings for Global optimization.
+func DefaultSettingsGlobal() *Settings {
+	return &Settings{
+		FunctionThreshold: math.Inf(-1),
+		FunctionConverge: &FunctionConverge{
+			Absolute:   1e-10,
+			Iterations: 100,
+		},
+	}
+}
 
-// GlobalMethod is an optimization method which seeks to find the global minimum
-// of an objective function.
-//
-// At the beginning of the optimization, InitGlobal is called to communicate
-// the dimension of the input and maximum number of concurrent tasks.
-// The actual number of concurrent tasks will be set from the return of InitGlobal,
-// which must not be greater than the input tasks.
-//
-// During the optimization, a reverse-communication interface is used between
-// the GlobalMethod and the caller.
-// GlobalMethod acts as a client that asks the caller to perform
-// needed operations given the return from IterateGlobal.
-// This allows and enforces automation of maintaining statistics and checking for
-// (various types of) convergence.
-//
-// The return from IterateGlobal can be an Evaluation, a MajorIteration or NoOperation.
-//
-// An evaluation is one or more of the Evaluation operations (FuncEvaluation,
-// GradEvaluation, etc.) combined with the bitwise or operator. In an evaluation
-// operation, the requested fields of Problem will be evaluated at the value
-// in Location.X, filling the corresponding fields of Location. These values
-// can be retrieved and used upon the next call to IterateGlobal with that task id.
-// The GlobalMethod interface requires that entries of Location are not modified
-// aside from the commanded evaluations. Thus, the type implementing GlobalMethod
-// may use multiple Operations to set the Location fields at a particular x value.
-//
-// When IterateGlobal declares MajorIteration, the caller updates the optimal
-// location to the values in Location, and checks for convergence. The type
-// implementing GlobalMethod must make sure that the fields of Location are valid
-// and consistent.
-//
-// IterateGlobal must not return InitIteration and PostIteration operations. These are
-// reserved for the clients to be passed to Recorders. A Method must also not
-// combine the Evaluation operations with the Iteration operations.
+// GlobalTask is a task to communicate between the GlobalMethod and
+type GlobalTask struct {
+	Index int
+	Operation
+	*Location
```

Comment:
Changed `Operation` to be a field rather than a method. `Location` doesn't have methods, but it does have fields, and so this allows writing `task.X` instead of `task.Loc.X`. I can change Location if you want too, it just seemed clearer this way.